Add TORCH_TARGET_VERSION for stable ABI

mikaylagawarecki opened the pull request:
And update it so comparisons can be done by the preprocessor

**Note: We also need to gate in shim.h and figure out how to enforce this**

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #165285
* #165284
* #165283
* #164332
* #163683
* #166373
* __->__ #164356

Differential Revision: [D85683549](https://our.internmc.facebook.com/intern/diff/D85683549)

mikaylagawarecki committed:
Add TORCH_TARGET_VERSION for stable ABI

[ghstack-poisoned]

janeyx99 commented on torch/csrc/stable/version.h:
```diff
@@ -0,0 +1,23 @@
+//     #include <torch/csrc/stable/library.h>
+//
+//   Target a specific stable version (major.minor):
+//     #define TORCH_TARGET_VERSION (((0ULL + 2) << 56) | ((0ULL + 9) << 48))
+//     Target PyTorch 2.9 #include <torch/csrc/stable/library.h>
```
```suggestion
//     #define TORCH_TARGET_VERSION (((0ULL + 2) << 56) | ((0ULL + 9) << 48))  // Target PyTorch 2.9
//     #include <torch/csrc/stable/library.h>
```
did lint mess this up?

janeyx99 commented on torch/csrc/stable/version.h:
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <torch/version.h>
```
I do think it'd be good to move this file to torch/headeronly to maintain our invariant

mikaylagawarecki replied:
opened https://github.com/pytorch/pytorch/pull/164381

mikaylagawarecki committed:
Update on "Add TORCH_TARGET_VERSION for stable ABI"

And update it so comparisons can be done by the preprocessor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add TORCH_TARGET_VERSION for stable ABI"

And update it so comparisons can be done by the preprocessor




[ghstack-poisoned]

mikaylagawarecki commented on torch/csrc/stable/version.h:
```diff
@@ -0,0 +1,23 @@
+//   Default behavior (uses current ABI version):
+//     #include <torch/csrc/stable/library.h>
+//
+//   Target a specific stable version (major.minor):
```
Ah I documented this in https://github.com/pytorch/pytorch/pull/163991

mikaylagawarecki committed:
Update on "Add TORCH_TARGET_VERSION for stable ABI"

And update it so comparisons can be done by the preprocessor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add TORCH_TARGET_VERSION for stable ABI"

And update it so comparisons can be done by the preprocessor

**Note: We also need to gate in shim.h and figure out how to enforce this**




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add TORCH_TARGET_VERSION for stable ABI"

And update it so comparisons can be done by the preprocessor

**Note: We also need to gate in shim.h and figure out how to enforce this**




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add TORCH_TARGET_VERSION for stable ABI"

And update it so comparisons can be done by the preprocessor

**Note: We also need to gate in shim.h and figure out how to enforce this**




[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D85638289).

mikaylagawarecki committed:
Update on "Add TORCH_TARGET_VERSION for stable ABI"

And update it so comparisons can be done by the preprocessor

**Note: We also need to gate in shim.h and figure out how to enforce this**


Differential Revision: [D85638289](https://our.internmc.facebook.com/intern/diff/D85638289)

[ghstack-poisoned]
